File: problems/chapter11/nicetak/arc090_b.jl
struct WeightedUnionFind
    par::Vector{Int}
    rank::Vector{Int}
    diff_weight::Vector{Number}

    WeightedUnionFind(n) = new(collect(1:n), zeros(Int, n), zeros(Number, n))
end

function root!(wuf::WeightedUnionFind, x::Int)
    if wuf.par[x] == x
        return x
    else
        r = root!(wuf, wuf.par[x])
        wuf.diff_weight[x] += wuf.diff_weight[wuf.par[x]]
        return wuf.par[x] = r
    end
end

function weight(wuf::WeightedUnionFind, x)
    root!(wuf, x) # Calculate weight
    return wuf.diff_weight[x]
end

diff(wuf::WeightedUnionFind, x::Int, y::Int) = weight(wuf, y) - weight(wuf, x)
issame(wuf::WeightedUnionFind, x::Int, y::Int) = root!(wuf, x) == root!(wuf, y)

function union!(wuf, x::Int, y::Int, w::Number)
    w += weight(wuf, x)    
    w -= weight(wuf, y)

    x = root!(wuf, x)
    y = root!(wuf, y)
    x == y && return false
    
    if wuf.rank[x] < wuf.rank[y]
        x, y = y, x
        w = -w
    elseif wuf.rank[x] == wuf.rank[y]
        wuf.rank[x] += 1
    end

    wuf.par[y] = x

    wuf.diff_weight[y] = w

    return true
end


function main()
    N, M = parse.(Int, split(readline()))

    wuf = WeightedUnionFind(N)
    ans = "Yes"
    for _ in 1:M
        L, R, D = parse.(Int, split(readline()))

        if issame(wuf, L, R)
            d = diff(wuf, L, R)
            if diff(wuf, L, R) ≠ D
                ans = "No"
                break
            end
        else
            union!(wuf, L, R, D)
        end
    end

    println(ans)
end

main()

File: problems/chapter11/Ysi/arc090_b.py
class WeightedUnionFind():
    def __init__(self, n) -> None:
        self.par = [-1] * n
        self.siz = [1] * n
        self.dist = [0] * n

    def root(self, x):
        if self.par[x] == -1:
            return x
        else:
            rootx = self.root(self.par[x])
            self.dist[x] += self.dist[self.par[x]]
            self.par[x] = rootx
            return self.par[x]

    def issame(self, x, y):
        return self.root(x) == self.root(y)

    def diff(self, x, y):
        if self.issame(x, y):
            return self.dist[y] - self.dist[x]
        else:
            return None

    def unite(self, x, y, w):
        w += self.dist[x] - self.dist[y]
        x = self.root(x)
        y = self.root(y)

        if x == y:
            return False

        if self.siz[x] < self.siz[y]:
            x, y, w = y, x, -w

        self.par[y] = x
        self.siz[x] += self.siz[y]
        self.dist[y] = w
        return True

    def size(self, x):
        return self.siz[self.root(x)]
def main():
    N, M = map(int, input().split())
    info = []
    for _ in range(M):
        L, R, D = map(int, input().split())
        info.append([L-1, R-1, D])

    wuf = WeightedUnionFind(N)
    ans = "Yes"
    for i in range(M):
        L, R, D = info[i]
        if wuf.diff(L, R) == None:
            wuf.unite(L, R, D)
        elif wuf.diff(L, R) != D:
            ans = "No"
            break
    
    print(ans)
if __name__=='__main__':
    main()